source/common/upstream/od_cds_api_impl.cc

```diff
@@ -0,0 +1,101 @@
+}
+
+OdCdsApiImpl::OdCdsApiImpl(const envoy::config::core::v3::ConfigSource& odcds_config,
+                           OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                           ClusterManager& cm, Stats::Scope& scope,
+                           ProtobufMessage::ValidationVisitor& validation_visitor)
+    : Envoy::Config::SubscriptionBase<envoy::config::cluster::v3::Cluster>(
+          odcds_config.resource_api_version(), validation_visitor, "name"),
+      helper_(cm, "odcds"), cm_(cm), scope_(scope.createScope("odcds.")),
+      status_(StartStatus::NotStarted) {
+  // TODO(krnowak): Move the subscription setup to CdsApiHelper. Maybe make CdsApiHelper a base
+  // class for CDS and ODCDS.
+  const auto resource_name = getResourceName();
+  if (!odcds_resources_locator.has_value()) {
+    subscription_ = cm_.subscriptionFactory().subscriptionFromConfigSource(
+        odcds_config, Grpc::Common::typeUrl(resource_name), *scope_, *this, resource_decoder_, {});
+  } else {
+    subscription_ = cm.subscriptionFactory().collectionSubscriptionFromUrl(
+        *odcds_resources_locator, odcds_config, resource_name, *scope_, *this, resource_decoder_);
+  }
+}
+
+void OdCdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& resources,
+                                  const std::string& version_info) {
+  UNREFERENCED_PARAMETER(resources);
+  UNREFERENCED_PARAMETER(version_info);
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+void OdCdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& added_resources,
+                                  const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                                  const std::string& system_version_info) {
+  auto exception_msgs =
+      helper_.onConfigUpdate(added_resources, removed_resources, system_version_info);
+  sendAwaiting();
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(
+        fmt::format("Error adding/updating cluster(s) {}", absl::StrJoin(exception_msgs, ", ")));
+  }
+}
+
+void OdCdsApiImpl::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
+                                        const EnvoyException*) {
+  ASSERT(Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure != reason);
+  sendAwaiting();
+}
+
+void OdCdsApiImpl::sendAwaiting() {
+  if (status_ != StartStatus::Started) {
+    return;
+  }
+  status_ = StartStatus::InitialFetchDone;
+  if (awaiting_names_.empty()) {
+    return;
+  }
+  ENVOY_LOG(debug, "odcds: sending request for awaiting cluster names {}",
+            fmt::join(awaiting_names_, ", "));
+  subscription_->requestOnDemandUpdate(awaiting_names_);
+  awaiting_names_.clear();
+}
+
+void OdCdsApiImpl::updateOnDemand(const std::string& cluster_name) {
+  switch (status_) {
+  case StartStatus::NotStarted:
+    ENVOY_LOG(trace, "odcds: starting a subscription with cluster name {}", cluster_name);
```

Comment:
An implementation question: why did you decide to delay the subscription start until an on-demand update request arrives? This is potentially expensive/time consuming if on-demand cds supports config server pushing a non-empty initial response.

Reply:
> An implementation question: why did you decide to delay the subscription start until an on-demand update request arrives?

My initial reason was to create a connection to config server only when it's actually needed. Second reason: the way I wanted ODCDS to work is that the config server sends responses that will only contain resources specifically asked for. And usually the empty initial request means "send me everything", right? Which is not what I wanted from config server when using ODCDS. From the config server point of view, if it receives an initial request from the stream (so, not asking for a specific resource and with an empty response nonce), how can it tell that this stream is from ODCDS or from CDS? Because for ODCDS a response would likely be empty, for CDS a response would be "all the clusters". So my idea here was to send an initial request that asks for a specific cluster. That way, when the config server receives a request without response nonce then checking whether it is a request for a specific resource (cluster) or not, it can decide whether to send this requested resource or everything. I think that this kind of logic would even work with ADS, if we configure ODCDS to use ADS.

This feels hacky, I admit. VHDS and probably SRDS don't have this problem, because these ask for different resource types (virtual hosts and scoped route configs, respectively), so config server could tell that requests for those resources are "on-demand", while ODCDS asks for the same resource types as CDS. Maybe there should be a different resource type for ODCDS?

> This is potentially expensive/time consuming if on-demand cds supports config server pushing a non-empty initial response.

I'm not sure I fully understand you here, but as I said above - I wouldn't want config server sending any special "initial response" to ODCDS. Which is why I make sure that the first request sent to config server already asks for a specific resource, and not for everything.

We have recently made a demo, where we demonstrated ODCDS with a super-simple config server: https://github.com/kinvolk/test-odcds/blob/master/server.go

Reply:
> I'm not sure I fully understand you here

The idea is that the config server can per-emptively send clusters if they can be useful; for example, other Envoy instances have recently made on-demand requests for these names (this is VHDS' behaviour). I don't know if want or should be consistent about these things, I'll let others chime in.